subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/PendingOptionalDependencies.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.api.internal.artifacts.ivyservice.resolveengine.graph.builder;
+
+import com.google.common.collect.Sets;
+
+import java.util.Set;
+
+public class PendingOptionalDependencies {
+    static final PendingOptionalDependencies NOT_OPTIONAL = new PendingOptionalDependencies(null);
+
+    private boolean noLongerOptional;
+    private final Set<NodeState> affectedComponents;
+
+    public PendingOptionalDependencies() {
+        this(Sets.<NodeState>newLinkedHashSet());
+    }
+
+    public PendingOptionalDependencies(Set<NodeState> nodeStates) {
```

Comment:
❌ I had to look three times at this class to make sure this was safe - a `null` `List` is always weird. Why not have a single private constructor with two arguments and two public factory methods?